### ISSUE TITLE
Custom touch button - configure small delay?

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#17251]
Got this through e-mail:

> In FIFA-14 ----- For "Advanced Offense" keypad setting---- (Classic Digital Dribble(Only Digital))
> 1. CHIP SHOT                         =  Circle + L
> 2. FINESSE SHOT                  =  L         + Circle
> Buttons --- Hence order of Button press becomes important
> The Program is always taking 1. for Custom Key setting as there is no order of Preference Option given---- Can this be implemented into the application ---- otherwise ---- FINESSE SHOT is not posssible!----This Is very Important

So basically, you can already config so that single touch buttons will trigger combinations of multiple PSP buttons, but they'll all key-down at exactly the same time. If we added a small delay between the first and second button, it would be possible to separately configure these two commands on different touch buttons.

I guess the question is, could we simply always apply a 1-frame delay on these combos between the first and second button, or would that break something else?

**Comment by unknownbrackets** [comment]
I feel like that's more of a "macro" concept.  For example, maybe I configure a combo to press R + triangle, because maybe that triggers scope aim or something.  I may have not thought about the order before and actually put triangle + R, but now you press triangle, then R.  Now the menu comes up, and R does nothing.

It may also be that the game wouldn't react to a single frame delay between these buttons, especially depending on how you define a frame.  Is that a vblank (games may read ctrl per vblank even if they are 30 FPS), or a drawn frame?

I think going down this road ends with defining macros with explicit delays.  In theory, this could be defined as:

> Step: Press L
> Step: Delay 100ms / 3 frames / whatever
> Step: Press Circle
>
> [　] Repeat macro while pressed
> [Ｘ] Finish macro even if released before complete

Though, that seems somewhat complicated.

-[Unknown]

**Comment by iota97** [comment]
I think your idea of making it always so it's more about making the UI simpler, but truth to be told we have no concept of buttons order in the UI and neither in the whole code.

I kinda find this complexity out of the scope of custom touch button. For this we will really need a macro system or some Lua scripting, or any other sane language to not anger Unknown with 1 indexed arrays :P

We could add to each custom button a drop down list with all possibile loaded script/macro from an ini in theory, still a totally different road.

**Comment by unknownbrackets** [comment]
Haha, it's just that I feel it's unnecessarily limiting and is less beneficial to users.
* If you have two people, and one knows a basic level of JavaScript or Python and the other knows a basic level of Lua, the former will be able to get more money.  So it's a more valuable skill, and game hacking is a great way to learn it.
* Learning languages with more standard indexing and etc. makes it easier to learn other languages too.
* I would challenge any assertion that Lua is easier to learn or has more widespread tutorials and help available for it than JavaScript or Python.
* They're both reasonably fast depending on what you're doing with them.

But I also think asking someone to dive into a scripting language so they can automate a two-button press in FIFA 14 seems unfortunate... although, there are external apps that may accomplish this too.

-[Unknown]

**Comment by iota97** [comment]
Yeah that's a good point, but we all know that when hrydgard will get to implement this he will go all the way to avoid possible bug reports from users making silly mistakes in scripts... And so we will have Rust scripting within PPSSPP.

More seriously I'm not so sure there are many app to add touch controls macro on Android, I might be wrong tho' :)

**Comment by ANR2ME** [comment]
Games that use ordered combo buttons usually use input events isn't? 
While games that doesn't care about the order can just check all the pressed buttons at once.

For example, when Circle pressed-down the games checked whether the firstButton(can be an array if there are more than 2 buttons sequence) has been set or not, if not set it to Circle.
When L being pressed the game checked whether the firstButton has been set or not, if it was set to Circle and matched to a combo sequence (ie.  Finesse Shot) then it do the combo action.
When all combo buttons has been released the game reset the firstButton (or clear the pressed buttons array)
